### PR TITLE
lsfd: report the signal masks of signalfds

### PR DESCRIPTION
**Pull request by masatake**
An example output:
    
        # ./lsfd -p "$(pidof systemd-journald)" -Q '(TYPE == "signalfd")'
    
        COMMAND             PID USER ASSOC MODE     TYPE       SOURCE MNTID INODE NAME
        systemd-journal 2382709 root   238  rw- signalfd anon_inodefs    15  1060 mask=USR1,USR2
        systemd-journal 2382709 root   239  rw- signalfd anon_inodefs    15  1060 mask=INT,TERM
        systemd-journal 2382709 root   240  rw- signalfd anon_inodefs    15  1060 mask=35

### CODE REVIEW

**Comment by masatake** [comment]
```
# ./lsfd -n -Q '(SOURCE == "anon_inodefs")' -o TYPE | sort | uniq -c | sort -nr
   1082   eventfd // done
   1081   timerfd // done
    426 eventpoll // done
    171   inotify 
     81     pidfd // done
     45  signalfd // this PR
     22  bpf-prog
      5   bpf-map
      2  fanotify
      1       btf
      1  bpf_link
$ grep -h -e anon_inode_getfd -r ../linux/[a-zA-Z]* | grep \"
		ret = anon_inode_getfd("kvm-spapr-tce", &kvm_spapr_tce_fops,
	ret = anon_inode_getfd("kvm-htab", &kvm_htab_fops, ctx, rwflag | O_CLOEXEC);
	fd = anon_inode_getfd("tee_shm", &tee_shm_fops, shm, O_RDWR);
	stream_fd = anon_inode_getfd("[i915_perf]", &fops, stream, f_flags);
	fd = anon_inode_getfd("iio:buffer", &iio_buffer_chrdev_fileops,
	fd = anon_inode_getfd("iio:event", &iio_event_chrdev_fileops,
	ufd = anon_inode_getfd("[timerfd]", &timerfd_fops, ctx,
	fd = anon_inode_getfd("[fscontext]", &fscontext_fops, fc,
		ufd = anon_inode_getfd("[signalfd]", &signalfd_fops, ctx,
	fd = anon_inode_getfd_secure("[userfaultfd]", &userfaultfd_fops, new,
	fd = anon_inode_getfd_secure("[userfaultfd]", &userfaultfd_fops, ctx,
	fd = anon_inode_getfd("[fanotify]", &fanotify_fops, group, f_flags);
	ret = anon_inode_getfd("inotify", &inotify_fops, group,
	return anon_inode_getfd("bpf-map", &bpf_map_fops, map,
	return anon_inode_getfd("bpf-prog", &bpf_prog_fops, prog,
	return anon_inode_getfd("bpf-link", &bpf_link_fops, link, O_CLOEXEC);
	fd = anon_inode_getfd("bpf-stats", &bpf_stats_fops, NULL, O_CLOEXEC);
	return anon_inode_getfd("btf", &btf_fops, btf, O_RDONLY | O_CLOEXEC);
	fd = anon_inode_getfd("[pidfd]", &pidfd_fops, get_pid(pid),
	ruleset_fd = anon_inode_getfd("[landlock-ruleset]", &ruleset_fops,
```

**Comment by masatake** [comment]
```
misc-utils/lsfd-unkn.c:763:26: error: implicit declaration of function 'sigabbrev_np' [-Werror=implicit-function-declaration]
    const char *signame = sigabbrev_np(signum);
                          ^~~~~~~~~~~~
misc-utils/lsfd-unkn.c:763:26: error: nested extern declaration of 'sigabbrev_np' [-Werror=nested-externs]
misc-utils/lsfd-unkn.c:763:26: error: initialization makes pointer from integer without a cast [-Werror=int-conversion]
```

Sigh.

File: misc-utils/lsfd-unkn.c [hunk 1]
```diff
@@ -715,6 +716,104 @@
...
+{
+	struct anon_signalfd_data *data = (struct anon_signalfd_data *)unkn->anon_data;
+
+	switch(column_id) {
+	case COL_SIGNALFD_MASK:
+		*str = anon_signalfd_make_mask_string(NULL, data->sigmask);
+		return true;
+	default:
+		return false;
+	}
```
[the remark this answers is not in the document]
Comment (karelzak) [reply misc-utils/lsfd-unkn.c:804]: Dear `Code scanning / CodeQL`, you're wrong.

**Comment by t-8ch** [comment]
> So I would like to decode the signal names only if sigabbrev_np is available on the platform.

Seems reasonable.

An alternative could be to do something like `tools/all_syscalls` is doing just for signal names.

File: misc-utils/lsfd-unkn.c [hunk 2]
```diff
@@ -715,6 +717,126 @@ static const struct anon_ops anon_timerfd_ops = {
...
+static_assert(sizeof(signals) > 0, "no signals found");
+
+static const char* ul_sigabbrev(int signum)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(signals); i++) {
+		if (signals[i].number == signum)
+			return signals[i].name;
+	}
+	return NULL;
+}
```
Comment (karelzak) [inline misc-utils/lsfd-unkn.c:743]: lib/signames.c: signum_to_signame()

Comment (masatake) [reply misc-utils/lsfd-unkn.c]: Thank you. Updated.

**Comment by masatake** [comment]
s390x and riscv64 reported the same error:

```
           lsfd: signalfd                       ...
  diff-{{{
  --- /home/runner/work/util-linux/util-linux/tests/expected/lsfd/mkfds-signalfd	2023-06-08 10:20:13.520555037 +0000
  +++ /home/runner/work/util-linux/util-linux/tests/output/lsfd/mkfds-signalfd	2023-06-08 11:12:21.150054369 +0000
  @@ -1,2 +1,2 @@
  -    3 signalfd mask=FPE,USR1,42   FPE,USR1,42
  +    3 signalfd mask=FPE,USR1,44   FPE,USR1,44
   ASSOC,TYPE,NAME,SIGNALFD.MASK: 0
  }}}-diff
  
   FAILED (lsfd/mkfds-signalfd)
```

s390x is big-endian. riscv64 is little-endian. So the failure is nothing to do with the endian.
So Qemu is suspicious.

I must find the way to run `build (qemu-user,s390x)` on my local pc.

**Pull request merged**